File: Lab3/C/labo3_ej1c.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>


void *calcular_sumatoria(void *arg){
	int *n = (int *) arg;
	int i;
	int sum = 0;
	for(i = 0; i < *n ; i++) {
		sum += i;
	}
	*n = sum;	
	pthread_exit(0);
}

int main(int argc, char *argv[]){

	if(argc < 2) {
		printf("Error en cantidad de argumentos\n");
		exit(1);
	}

	int cantHilos = atoi(argv[1]);
	pthread_t hilos[cantHilos];
	int i;
	int retornoCreate;
	int sumatorias[cantHilos];

	for(i = 0; i < cantHilos; i++){
		sumatorias[i] = i + cantHilos;
		retornoCreate =	pthread_create(&hilos[i], NULL, calcular_sumatoria, (void *) &sumatorias[i]);
		if(retornoCreate != 0) {
			printf("Error creando hilo\n");
			exit(1);
		}
	}

	for(i = 0; i < cantHilos; i++){
		pthread_join(hilos[i], NULL);
		printf("Termino hilo %d con sumatoria(de %d a %d): %d\n", (i+1), 0, ((i + cantHilos) - 1), sumatorias[i]);
	}

	return 0;
}

File: Lab3/Rust/labo3_ej1c.rs
use std::env;
use std::thread;

fn calcular_sum(n: usize) -> usize {
    let mut sum: usize = 0;
    for i in 0..n {
        sum = sum + i;
    }
    sum
}


fn main() {
    let args: Vec<String> = env::args().collect();

    if args.len() != 2 {
        println!("Error en argumentos por linea de comando");
        std::process::exit(1);
    }

    let n: usize = args[1].parse().expect("Entero");
    let mut vector_hilos = Vec::with_capacity(n);

    for i in 0..n {
        let tope_sum = i + n;
        let hilo = thread::spawn(move || {
            calcular_sum(tope_sum)
        });
        vector_hilos.push(hilo);
    }   
    let mut i: usize = 0;
    for hilo in vector_hilos {
        let resultado = hilo.join().unwrap();
        println!("Termino hilo {} con sumatoria(de {} a {}): {:?}", (i+1), 0, (i), resultado);
        i += 1;
    }
}
